### PR TITLE
Get better OpenGL version than 2.1 on macOS

### PR DESCRIPTION
**Pull request by metayan**
Normally macOS gives us only OpenGL 2.1, but if we request 3.3 and create a new context, we get the best OpenGL version it has to offer.

Before:
INFO: Renderer: opengl
INFO: OpenGL version: 2.1 NVIDIA-14.0.32 355.11.11.10.10.143
WARN: Trilinear filtering disabled (OpenGL 3.0+ or ES 2.0+ required)

After:
INFO: Renderer: opengl
DEBUG: Creating GL Context
INFO: OpenGL version: 4.1 NVIDIA-14.0.32 355.11.11.10.10.143
INFO: Trilinear filtering enabled

when running with
--verbosity=verbose --render-driver=opengl

Note:
Since SDL_CreateRenderer causes a fallback to OpenGL 2.1, the profile and version attributes have to be set and the context created _after_.
This took a while to figure out.

### CODE REVIEW

**Comment by metayan** [comment]
> > Since SDL_CreateRenderer causes a fallback to OpenGL 2.1
> 
> Do you know where (in the SDL codebase)?

Found it mentioned at
https://stackoverflow.com/a/23858884
"… SDL lowering the context version is due to the fact that the renderer is implemented with OpenGL 2.1 calls, so it needs to create a 2.1 GL context."

Also discussed at
https://github.com/libsdl-org/SDL/issues/1957
which ends with
"Marking won't fix for now, since the SDL render API doesn't support Core Profile."


> > if we request 3.3 and create a new context, we get the best OpenGL version it has to offer.
> 
> What if the latest version available is <3.3? It will fail in that case.

Changed to select OpenGL 3.2.

According to page 24 of
https://developer.apple.com/opengl/OpenGL-Capabilities-Tables.pdf
all hardware and software renderers can use OpenGL 3.2 since macOS 10.7.5.

Also, when choosing 3.2, I actually get 4.1, which is the highest that my computer supports under macOS:
```
INFO: Renderer: opengl
DEBUG: Creating OpenGL 3.2 Context
INFO: OpenGL version: 4.1 NVIDIA-14.0.32 355.11.11.10.10.143
INFO: Trilinear filtering enabled
```
Apparently that is how Apple chose to handle it:
"for macOS you either have 3.2 or 4.1, nothing in between"
from https://github.com/ocornut/imgui/issues/1466#issuecomment-352058971

A more thorough explanation of macOS OpenGL peculiarities:
https://github.com/ocornut/imgui/issues/1466#issuecomment-354128340


> I'm a bit afraid that this change is specific to your machine/drivers, and that it may fail otherwise.

It seems like it's specific to OpenGL under macOS in general.
For the record, I'm using Catalina (10.15.7).
The change only affects the OpenGL drivers, so those who don't specify `--render-driver=opengl` get METAL under macOS.


> If OpenGL 3 or 4 can be available on macOS, IMO it should even be changed on SDL directly (we can still try to make it work better on scrcpy temporarily).

The option to use higher OpenGL versions under macOS has been available for a long time, and SDL supports using them, but not automagically.

A solution in Lisp, for example, which starts with the highest version and keeps decreasing it until a context is created:
https://github.com/cbaggers/cepl.sdl2/blob/6da5a030db5e3579c5a1c5350b1ffb8fc9950e9a/cepl.sdl2.lisp#L139-L155

File: app/src/screen.c [hunk 1]
```diff
@@ -470,6 +470,21 @@ sc_screen_init(struct sc_screen *screen,
     // starts with "opengl"
     bool use_opengl = renderer_name && !strncmp(renderer_name, "opengl", 6);
     if (use_opengl) {
+
+#ifdef __APPLE__
+        // Persuade macOS to give us something better than OpenGL 2.1
+        SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+        SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
+        SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3);
```
[the remark this answers is not in the document]
Comment (metayan) [reply app/src/screen.c]: It's a macOS OpenGL strangeness that SDL seems to have left up to us to solve.
More explained in https://github.com/Genymobile/scrcpy/pull/3895#issuecomment-1498297668 below.

File: app/src/screen.c [hunk 2]
```diff
@@ -470,6 +470,21 @@ sc_screen_init(struct sc_screen *screen,
...
+        LOGD("Creating GL Context");
+        SDL_GLContext *gl_context;
+        gl_context = SDL_GL_CreateContext(screen->window);
```
[the remark this answers is not in the document]
Comment (metayan) [reply app/src/screen.c]: Yes. Added `SDL_GL_DeleteContext()`.

**Comment by metayan** [comment]
By only setting the Core Profile for the context, without specifying the OpenGL version, macOS seems to choose the highest available OpenGL version.

I've modified the PR accordingly.

The reason why SDL doesn't do this automagically:

> Platform Issue (MacOSX): When MacOSX 10.7 introduced support for OpenGL beyond 2.1, they also introduced the core/compatibility dichotomy. However, they did not introduce support for the compatibility profile itself. Instead, MacOSX gives you a choice: core profile for versions 3.2 or higher, or just version 2.1. There is no way to get access to features after 2.1 and still access the [Fixed Function Pipeline](https://www.khronos.org/opengl/wiki/Fixed_Function_Pipeline).

From https://www.khronos.org/opengl/wiki/OpenGL_Context#OpenGL_3.2_and_Profiles

**Comment by rom1v** [comment]
Thank you for the details :+1:

I had a refactor commit related to screen/display on a local branch that I just pushed to `dev`: 051b74c88338702508d98bd532057b343e0a177c

I rebased/reapplied your work above. Please check and review branch [`macos_opengl`](https://github.com/Genymobile/scrcpy/commits/macos_opengl).

**Comment by metayan** [comment]
Looks good and works now.

**Comment by rom1v** [comment]
You should not have closed it, it's not merged yet, and I cannot reopen since your branch has been deleted :/

I would like other people to test on macOS before merging :)

**Comment by metayan** [comment]
Oh, sorry.
Luckily I hadn't deleted it locally yet, so pushed it again.
Reopening. Hope it works.

**Comment by metayan** [comment]
I thought that it's fine to delete, since it's in `macos_opengl`.
Well, we learn... ;)

**Comment by metayan** [comment]
Should I do something about the "This branch has conflicts that must be resolved" that's being shown at the end of this thread?

**Comment by rom1v** [comment]
Merged :+1: Thank you.

> I would like other people to test branch [`macos_opengl`](https://github.com/Genymobile/scrcpy/commits/macos_opengl) (with `scrcpy --Vdebug --render-driver=opengl`) on macOS before merging :smile: 

I had a confirmation that it works as expected. :+1:

**Pull request closed without merging**